// @sanity/preview-url-secret/dist/constants.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const schemaType = "sanity.previewUrlSecret";
const schemaIdPrefix = "sanity-preview-url-secret";
const schemaIdSingleton = `${schemaIdPrefix}.share-access`;
const schemaTypeSingleton = "sanity.previewUrlShareAccess";
const apiVersion = "2023-11-09";
const urlSearchParamPreviewSecret = "sanity-preview-secret";
const urlSearchParamPreviewPathname = "sanity-preview-pathname";
const urlSearchParamPreviewPerspective = "sanity-preview-perspective";
const isDev = process.env.NODE_ENV === "development";
const SECRET_TTL = 60 * 60;
const fetchSecretQuery = (
  /* groq */
  `*[_type == "${schemaType}" && secret == $secret && dateTime(_updatedAt) > dateTime(now()) - ${SECRET_TTL}][0]{
    _id,
    _updatedAt,
    secret,
    studioUrl,
  }`
);
const fetchSharedAccessQuery = (
  /* groq */
  `*[_id == "${schemaIdSingleton}" && _type == "${schemaTypeSingleton}"][0].secret`
);
const fetchSharedAccessSecretQuery = (
  /* groq */
  `*[_id == "${schemaIdSingleton}" && _type == "${schemaTypeSingleton}" && secret == $secret][0]{
  secret,
  studioUrl,
}`
);
const deleteExpiredSecretsQuery = (
  /* groq */
  `*[_type == "${schemaType}" && dateTime(_updatedAt) <= dateTime(now()) - ${SECRET_TTL}]`
);
const tag = "sanity.preview-url-secret";
const perspectiveCookieName = "sanity-preview-perspective";

exports.SECRET_TTL = SECRET_TTL;
exports.apiVersion = apiVersion;
exports.deleteExpiredSecretsQuery = deleteExpiredSecretsQuery;
exports.fetchSecretQuery = fetchSecretQuery;
exports.fetchSharedAccessQuery = fetchSharedAccessQuery;
exports.fetchSharedAccessSecretQuery = fetchSharedAccessSecretQuery;
exports.isDev = isDev;
exports.perspectiveCookieName = perspectiveCookieName;
exports.schemaIdPrefix = schemaIdPrefix;
exports.schemaIdSingleton = schemaIdSingleton;
exports.schemaType = schemaType;
exports.schemaTypeSingleton = schemaTypeSingleton;
exports.tag = tag;
exports.urlSearchParamPreviewPathname = urlSearchParamPreviewPathname;
exports.urlSearchParamPreviewPerspective = urlSearchParamPreviewPerspective;
exports.urlSearchParamPreviewSecret = urlSearchParamPreviewSecret;
