// sanity/lib/_legacy/ViteDevServerStopped.esm.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Card, Container, Stack, Heading, Text } from "@sanity/ui";
import { useState, useCallback, useEffect } from "react";
const ERROR_TITLE = "Dev server stopped", ERROR_DESCRIPTION = "The development server has stopped. You may need to restart it to continue working.";
class ViteDevServerStoppedError extends Error {
  ViteDevServerStoppedError;
  constructor() {
    super(ERROR_TITLE), this.name = "ViteDevServerStoppedError", this.ViteDevServerStoppedError = !0;
  }
}
const serverHot = import.meta.hot, isViteServer = (hot) => !!hot, useDetectViteDevServerStopped = () => {
  const [devServerStopped, setDevServerStopped] = useState(!1), markDevServerStopped = useCallback(() => setDevServerStopped(!0), []);
  return useEffect(() => {
    isViteServer(serverHot) && serverHot.on("vite:ws:disconnect", markDevServerStopped);
  }, [markDevServerStopped]), { devServerStopped };
}, ThrowViteServerStopped = () => {
  const { devServerStopped } = useDetectViteDevServerStopped();
  if (devServerStopped) throw new ViteDevServerStoppedError();
  return null;
}, DetectViteDevServerStopped = () => isViteServer(serverHot) ? /* @__PURE__ */ jsx(ThrowViteServerStopped, {}) : null, DevServerStoppedErrorScreen = () => /* @__PURE__ */ jsx(
  Card,
  {
    height: "fill",
    overflow: "auto",
    paddingY: [4, 5, 6, 7],
    paddingX: 4,
    sizing: "border",
    tone: "critical",
    children: /* @__PURE__ */ jsx(Container, { width: 3, children: /* @__PURE__ */ jsxs(Stack, { space: 4, children: [
      /* @__PURE__ */ jsx(Heading, { children: ERROR_TITLE }),
      /* @__PURE__ */ jsx(Card, { border: !0, radius: 2, overflow: "auto", padding: 4, tone: "inherit", children: /* @__PURE__ */ jsx(Stack, { space: 4, children: /* @__PURE__ */ jsx(Text, { size: 2, children: ERROR_DESCRIPTION }) }) })
    ] }) })
  }
);
export {
  DetectViteDevServerStopped,
  DevServerStoppedErrorScreen
};
