// sanity/lib/_chunks-cjs/ViteDevServerStopped.js
"use strict";
var jsxRuntime = require("react/jsx-runtime"), ui = require("@sanity/ui"), React = require("react");
const ERROR_TITLE = "Dev server stopped", ERROR_DESCRIPTION = "The development server has stopped. You may need to restart it to continue working.";
class ViteDevServerStoppedError extends Error {
  ViteDevServerStoppedError;
  constructor() {
    super(ERROR_TITLE), this.name = "ViteDevServerStoppedError", this.ViteDevServerStoppedError = !0;
  }
}
const serverHot = void 0, isViteServer = (hot) => !!hot, useDetectViteDevServerStopped = () => {
  const [devServerStopped, setDevServerStopped] = React.useState(!1), markDevServerStopped = React.useCallback(() => setDevServerStopped(!0), []);
  return React.useEffect(() => {
    isViteServer(serverHot) && serverHot.on("vite:ws:disconnect", markDevServerStopped);
  }, [markDevServerStopped]), { devServerStopped };
}, ThrowViteServerStopped = () => {
  const { devServerStopped } = useDetectViteDevServerStopped();
  if (devServerStopped) throw new ViteDevServerStoppedError();
  return null;
}, DetectViteDevServerStopped = () => isViteServer(serverHot) ? /* @__PURE__ */ jsxRuntime.jsx(ThrowViteServerStopped, {}) : null, DevServerStoppedErrorScreen = () => /* @__PURE__ */ jsxRuntime.jsx(
  ui.Card,
  {
    height: "fill",
    overflow: "auto",
    paddingY: [4, 5, 6, 7],
    paddingX: 4,
    sizing: "border",
    tone: "critical",
    children: /* @__PURE__ */ jsxRuntime.jsx(ui.Container, { width: 3, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { space: 4, children: [
      /* @__PURE__ */ jsxRuntime.jsx(ui.Heading, { children: ERROR_TITLE }),
      /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { border: !0, radius: 2, overflow: "auto", padding: 4, tone: "inherit", children: /* @__PURE__ */ jsxRuntime.jsx(ui.Stack, { space: 4, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: 2, children: ERROR_DESCRIPTION }) }) })
    ] }) })
  }
);
exports.DetectViteDevServerStopped = DetectViteDevServerStopped;
exports.DevServerStoppedErrorScreen = DevServerStoppedErrorScreen;
