// @sanity/presentation/dist/_chunks-cjs/PostMessageDocuments.cjs
"use strict";
var react = require("react"), rxjs = require("rxjs"), sanity = require("sanity"), index = require("./index.cjs");
const PostMessageDocuments = (props) => {
  const { comlink } = props, client = sanity.useClient({ apiVersion: index.API_VERSION });
  return react.useEffect(() => {
    const listener = client.listen(
      '*[!(_id in path("_.**"))]',
      {},
      {
        effectFormat: "mendoza",
        events: ["welcome", "mutation", "reconnect"],
        includePreviousRevision: !1,
        includeResult: !1,
        tag: "presentation-documents",
        visibility: "transaction"
      }
    ).pipe(
      rxjs.filter(
        (event) => event.type === "welcome" || event.type === "reconnect" || event.type === "mutation"
      )
    ), welcome = listener.pipe(
      rxjs.filter((event) => event.type === "welcome"),
      rxjs.shareReplay({ bufferSize: 1, refCount: !1 })
    ), unsubscribe = comlink.on("visual-editing/snapshot-welcome", async () => ({ event: await new Promise((resolve) => {
      welcome.pipe(rxjs.first()).subscribe((event2) => {
        resolve(event2);
      });
    }) })), reconnect = listener.pipe(
      rxjs.filter((event) => event.type === "reconnect")
    ), mutations = listener.pipe(
      rxjs.filter((event) => event.type === "mutation")
    ), events = rxjs.merge(
      /**
       * @deprecated remove 'welcome' here and switch to explict welcome message fetching at next major
       */
      welcome,
      mutations,
      reconnect
    ).subscribe((event) => {
      comlink.post({ type: "presentation/snapshot-event", data: { event } });
    });
    return () => {
      unsubscribe(), events.unsubscribe();
    };
  }, [client, comlink]), react.useEffect(() => comlink.on("visual-editing/fetch-snapshot", async (data) => ({ snapshot: await client.getDocument(data.documentId, {
    tag: "document.snapshots"
  }) })), [client, comlink]), react.useEffect(() => comlink.on("visual-editing/mutate", async (data) => client.dataRequest("mutate", data, {
    visibility: "async",
    returnDocuments: !0
  })), [client, comlink]), null;
};
var PostMessageDocuments$1 = react.memo(PostMessageDocuments);
exports.default = PostMessageDocuments$1;
